usr/lib/common/pin_get.h

```diff
@@ -0,0 +1,86 @@
+/*
+ * COPYRIGHT (c) International Business Machines Corp. 2022
+ *
+ * This program is provided under the terms of the Common Public License,
+ * version 1.0 (CPL-1.0). Any use, reproduction or distribution for this
+ * software constitutes recipient's acceptance of CPL-1.0 terms which can be
+ * found in the file LICENSE file or at
+ * https://opensource.org/licenses/cpl1.0.php
+ */
+
+#ifndef _PIN_GET_H_
+#define _PIN_GET_H_
+
+struct pinbuf {
+    char *p;
+    size_t len;
+};
```

Comment:
Is there a good reason why you introduce this struct? Why not simply use a zero terminated string?
Its essentially a buffer containing a string with a specific length. 

The only good reason I can imagine is to allow pins to include the \0 character in its payload.
But then you can't use strlen, strncmp, etc to operate on the pin contents (which you currently do in your implementation). 

In my opinion this makes it unnecessarily complicated. Just use a zero terminated string instead, and trust that the string is in fact zero terminated.

Reply:
The problem for the pin_get_*() functions is, that we have 2-3 different source for the pin pointers, argv, env variable or the buffer from the input (getline()). Only the latter on is dynamic and only for this purpose, we need a char pointer and a size. Note: len is the length of allocated space in p, not the length of the input.

So, from a user point of view, the pin should be accessed only via the return value of the pin_get_*() or pin_prompt() functions. The stuct pinbuf variables has the only purpose, to keep track of the dynamically allocated memory. We do not really need a struct here, we could also just use a pointer to a char and a pointer to a size_t variable. But having a struct here, makes it a bit less complicated in my opinion.

BTW: I do not agree, that we should support \0 in pins. All strings, which are returned by pin_get_*() function, are null-terminated.

Reply:
> The problem for the pin_get_*() functions is, that we have 2-3 different source for the pin pointers, argv, env variable or the buffer from the input (getline()). Only the latter on is dynamic and only for this purpose, we need a char pointer and a size.

Yes, I understood that.  

> We do not really need a struct here, we could also just use a pointer to a char and a pointer to a size_t variable.

Do we really need the (allocated) size? Anywhere where we need the length, we can do `strlen()`. 
But yes, we do need the allocated pointer, e.g. to free it afterwards.

So instead of passing `struct pinbuf *buf` we could simply pass `char **buf` (or even `void **buf` if you like) and treat `buf` the same way as before, i.e. pass it to `pinbuf_free()`, `pin_get_user()`, etc.

Reply:
But if we remove the length information, we can not wipe out the pin on free. And we loose the possibility to work with pre-allocated buffers (because without a length, we do not know, how much space has been allocated). 

What about the following:
Make `struct pinbuf`` only visible to pin_get.c and use a `typedef pinbuf_t struct pinbuf` in the header file. So all users need a pinbuf_t variable to work with the pin_get functions. I used such a mechanism for the buffer.c as well.

Reply:
Hm... why not usig p11_buffer_t for that purpose? It has all what we need. Let's give it a try.

Reply:
> But if we remove the length information, we can not wipe out the pin on free.

Sure you can: just use `strlen()` to get the length to wipe. Wiping the terminating zero by another zero does not add much :-) 

> And we loose the possibility to work with pre-allocated buffers

Are you actually using this somewhere in the code? If not, I guess we don't need that possibility.

> Make struct pinbuf only visible to pin_get.c and use a typedef pinbuf_t struct pinbuf in the header file.

Thats better, but I still think you only need a `char*` and not a struct.

> Hm... why not using p11_buffer_t for that purpose? It has all what we need. Let's give it a try.

Again, I don't think all this is really needed. Lets please keep it as simple as possible.

Reply:
ok